tests/src/Interop/ArrayMarshalling/SafeArray/SafeArrayNative.cpp

```diff
@@ -0,0 +1,651 @@
+
+	// Get a pointer to the elements of the array.
+	hr = SafeArrayAccessData(*ppsa, (void **)&pInt);
+	if(FAILED(hr))
+	{
+		printf("\t\tSafeArrayAccessData call in SafeArray_InOutByRef failed!\n");
+		return FALSE;
+	}
+
+	//get upperbound
+	hr = SafeArrayGetUBound(*ppsa, 1, &lUbound);
+	if(FAILED(hr))
+	{
+		printf("\t\tSafeArrayGetUBound call in SafeArray_InOutByRef failed!\n");
+		return FALSE;
+	}
+
+	//check upperbound
+	if(lUbound != 255) //since num of elems = lUBound - lLBound + 1
+	{
+		printf("\t\tlUbound not as expected in SafeArray_InOutByRef!\n");
+		printf("\t\t\tlUbound = %d",lUbound);
+		return FALSE;
+	}
+
+	//get lowerbound
+	hr = SafeArrayGetLBound(*ppsa, 1, &lLbound);
+	if(FAILED(hr))
+	{
+		printf("\t\tSafeArrayGetLBound call in SafeArray_InOutByRef failed!\n");
+		return FALSE;
+	}
+
+	//check lowerbound
+	if(lLbound != 0) 
+	{
+		printf("\t\tlLbound not as expected in SafeArray_InOutByRef!\n");
+		printf("\t\t\tlLbound = %d",lLbound);
+		return FALSE;
+	}
+
+	//check dimension
+	if(SafeArrayGetDim(*ppsa) != 1)
+	{
+		printf("\t\tDimension not as expected in SafeArray_InOutByRef!\n");
+		return FALSE;
+	}
+
+	//check element size
+	if(SafeArrayGetElemsize(*ppsa) != 4) //size of each element should be 4 bytes
+	{
+		printf("\t\tElement size not as expected in SafeArray_InOut!\n");
+		return FALSE;
+	}
+
+	//validate data
+	for(i = 0; i <= lUbound; i++)
+	{
+		if(pInt[i] != i)
+		{
+			printf("\t\tData not as expected in SafeArray_InOut!\n");
+			return FALSE;
+		}
+	}
+
+	// Change the safearray 
+
+	//reverse data
+	for(i = 0; i <= lUbound; i++)	
+		pInt[i] = lUbound - i;
+
+	SafeArrayUnaccessData(*ppsa);
+	return TRUE;
+}
+
+extern "C" DLL_EXPORT BOOL STDMETHODCALLTYPE SafeArrayWithOutAttribute(SAFEARRAY* psa)
+{
+	long i, lUbound;
+	int  *pInt; //will point to the data
+	HRESULT hr;
+
+	// Get a pointer to the elements of the array.
+	hr = SafeArrayAccessData(psa, (void **)&pInt);
+	if(FAILED(hr))
+	{
+		printf("\t\tSafeArrayAccessData call in SafeArray_InOut failed!\n");
+		return FALSE;
+	}
+
+	//get upperbound
+	hr = SafeArrayGetUBound(psa, 1, &lUbound);
+	if(FAILED(hr))
+	{
+		printf("\t\tSafeArrayGetUBound call in SafeArray_InOut failed!\n");
+		return FALSE;
+	}
+
+	//check upperbound
+	if(lUbound != 255) //since num of elems = lUBound - lLBound + 1
+	{
+		printf("\t\tlUbound not as expected in SafeArray_InOut!\n");
+		printf("\t\t\tlUbound = %d",lUbound);
+		return FALSE;
+	}
+
+	//check dimension
+	if(SafeArrayGetDim(psa) != 1)
+	{
+		printf("\t\tDimension not as expected in SafeArray_InOut!\n");
+		return FALSE;
+	}
+
+	/************************************************
+	* Change the safearray 
+	************************************************/
+
+	//change the elements to -1
+	for(i = 0; i <= lUbound; i++)	
+		pInt[i] = -1;
+
+	SafeArrayUnaccessData(psa);
+	return TRUE;
+}
+
+extern "C" DLL_EXPORT SAFEARRAY* STDMETHODCALLTYPE SafeArray_Ret_MismatchRank()
+{
+	SAFEARRAY * psa;
+
+	const int rank = 2;
```

Comment:
Again, what rank is wrong? This is setting the rank to 2? Is that wrong? A nice indication of this is defining the variable as `const int InvalidRank = 2;` that would make it clear 2 isn't the valid value.